### PR TITLE
#11644: Migrate AutoFormat to TTNN Experimental

### LINKED ISSUE #11644: Move Autoformat to ttnn
Master Issue : https://github.com/tenstorrent/tt-metal/issues/11513

### PR DESCRIPTION
**Pull request by mouliraj-mcw**
### Ticket
Link to Github Issue #11644 .

### What's changed

- Migrated the AutoFormat to TTNN Experimental
- Moved following apis from ttnn.experimental.__ to ttnn.__

1. pad_to_tile_shape
2. format_input_tensor
3. format_output_tensor
4. SetDefaultDevice
5. GetDefaultDevice
 
### Checklist
- [x] [All Post commit CI ](https://github.com/tenstorrent/tt-metal/actions/runs/10602730054)
- [x]  [(Single-card) Model perf tests](https://github.com/tenstorrent/tt-metal/actions/runs/10602763745)
- [x] [(Single-card) Demo tests](https://github.com/tenstorrent/tt-metal/actions/runs/10602744410)
- [x] [(Single-card) Device perf regressions](https://github.com/tenstorrent/tt-metal/actions/runs/10602753743)
- [x] [Nightly fast dispatch tests](https://github.com/tenstorrent/tt-metal/actions/runs/10602773033)
- [x] [[post-commit] models tests](https://github.com/tenstorrent/tt-metal/actions/runs/10602804234)
- [x] [(T3K) T3000 model perf tests](https://github.com/tenstorrent/tt-metal/actions/runs/10602823123)
- [x] [(T3K) T3000 frequent tests](https://github.com/tenstorrent/tt-metal/actions/runs/10602833889)
- [x] [(T3K) T3000 nightly tests](https://github.com/tenstorrent/tt-metal/actions/runs/10602845790)

### CODE REVIEW

File: conftest.py [hunk 1]
```diff
@@ -111,14 +111,15 @@ def device_params(request):
 @pytest.fixture(scope="function")
 def device(request, device_params):
     import tt_lib as ttl
```
Comment (umadevimcw) [inline conftest.py:113]: is ttl import required? if not remove it

File: models/experimental/llama/tests/test_llama_mlp.py [hunk 2]
```diff
@@ -97,7 +97,7 @@ def run_test_LlamaMLP_inference(device, model_version, tokenizer_version, batch,
 def test_LlamaMLP_inference(model_version, tokenizer_version, batch, seq_len, on_weka, pcc):
     # Initialize the device
     device = ttnn.open_device(0)
-    ttnn.experimental.device.SetDefaultDevice(device)
+    ttnn.SetDefaultDevice(device)
 
     run_test_LlamaMLP_inference(device, model_version, tokenizer_version, batch, seq_len, on_weka, pcc)
-    ttnn.close_device(device)
+    ttnn.CloseDevice(device)
```
Comment (KalaivaniMCW) [inline models/experimental/llama/tests/test_llama_mlp.py:103]: do we need to change this from close_device ?

Comment (mouliraj-mcw) [reply models/experimental/llama/tests/test_llama_mlp.py:103]: No. will revert it. Thankyou

File: ttnn/cpp/ttnn/operations/transformer/sdpa_decode_gqa/sdpa_decode_gqa.cpp [hunk 3]
```diff
@@ -35,7 +35,7 @@ ttnn::Tensor ExecuteScaledDotProductAttentionGQADecode::invoke(
     std::optional<SDPAProgramConfig> program_config,
     std::optional<DeviceComputeKernelConfig> compute_kernel_config) {
     auto arch = input_tensor_q.storage_type() == StorageType::DEVICE ? input_tensor_q.device()->arch()
-                                                                     : AutoFormat::GetDefaultDevice()->arch();
+                                                                     : ttnn::operations::experimental::auto_format::AutoFormat::GetDefaultDevice()->arch();
```
Comment (Aswinmcw) [inline ttnn/cpp/ttnn/operations/transformer/sdpa_decode_gqa/sdpa_decode_gqa.cpp:38]: Is full path needed? ttnn:: alone will not work?

File: ttnn/cpp/ttnn/operations/data_movement/concat/device/concat_device_operation.cpp [hunk 4]
```diff
@@ -7,11 +7,11 @@
 
 #include "ttnn/tensor/tensor.hpp"
 #include "ttnn/tensor/tensor_utils.hpp"
-#include "ttnn/deprecated/tt_dnn/op_library/auto_format.hpp"
+#include "ttnn/operations/experimental/auto_format/auto_format.hpp"
 #include "ttnn/run_operation.hpp"
 
 using namespace tt::constants;
-
+using ttnn::operations::experimental::auto_format::FormatParams;
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/data_movement/concat/device/concat_device_operation.cpp:14]: can this be moved inside concat_impl if that's the only fn using FormatParams?

File: ttnn/cpp/ttnn/operations/data_movement/permute/permute.cpp [hunk 5]
```diff
@@ -9,12 +9,13 @@
...
 #include "tt_metal/host_api.hpp"
 #include "tt_metal/common/constants.hpp"
-#include "ttnn/deprecated/tt_dnn/op_library/auto_format.hpp"
+#include "ttnn/operations/experimental/auto_format/auto_format.hpp"
 #include "ttnn/tensor/tensor_utils.hpp"
 
 #include "ttnn/operations/core/core.hpp"
 #include "ttnn/run_operation.hpp"
 
+using ttnn::operations::experimental::auto_format::AutoFormat;
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/data_movement/permute/permute.cpp:18]: same comment as earlier.. can this be moved to the respective fn using it?

File: ttnn/cpp/ttnn/operations/experimental/auto_format/auto_format.cpp [hunk 6]
```diff
@@ -234,5 +231,4 @@ Tensor AutoFormat::format_output_tensor(
     return formatted_output;
 }
 
-}  // namespace tt_metal
-}  // namespace tt
+}  // ttnn::operations::auto_format namespace
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/experimental/auto_format/auto_format.cpp:234]: // namespace ttnn::operations::auto_format

File: ttnn/cpp/pybind11/device.hpp [hunk 7]
```diff
@@ -30,6 +31,37 @@ void py_module(py::module& module) {
...
+        py::arg("target_layout").noconvert(),
+        py::arg("target_mem_config").noconvert() = std::nullopt);
+    module.def(
+        "pad_to_tile_shape",
+        [](const std::array<uint32_t, 4>& unpadded_shape,
+           bool pad_c = false,
+           bool pad_n = false,
+           bool pad_h = true,
+           bool pad_w = true) -> Shape {
+            return Shape(ttnn::operations::experimental::auto_format::AutoFormat::pad_to_tile_shape(unpadded_shape, pad_c, pad_n, pad_h, pad_w));
```
[the remark this answers is not in the document]
Comment (mouliraj-mcw) [reply ttnn/cpp/pybind11/device.hpp:62]: tt::tt_metal::Shape

[the remark this answers is not in the document]
Comment (KalaivaniMCW) [reply ttnn/cpp/pybind11/device.hpp:62]: mention it explicitly as tt::tt_metal::Shape, as this file is under ttnn and there is also another ttnn::Shape

File: ttnn/cpp/ttnn/operations/experimental/auto_format/auto_format.hpp [hunk 8]
```diff
@@ -3,22 +3,21 @@
 // SPDX-License-Identifier: Apache-2.0
 
 #pragma once
-#include "ttnn/tensor/tensor.hpp"
+#include "ttnn/tensor/types.hpp"
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/experimental/auto_format/auto_format.hpp:6]: is tensor.hpp not required here ? only types ?

**Review by KalaivaniMCW — APPROVED** [review]
Approved

**Review by bbradelTT — APPROVED** [review]

**Review by tt-aho — APPROVED** [review]

**Review by sankarmanoj-tt — APPROVED** [review]

File: conftest.py [hunk 9]
```diff
@@ -111,14 +111,15 @@ def device_params(request):
...
+    import ttnn
 
     device_id = request.config.getoption("device_id")
     request.node.pci_ids = [ttl.device.GetPCIeDeviceID(device_id)]
```
Comment (tt-rkim) [inline conftest.py:117]: Is there still not a function `ttnn` equivalent that does this?

File: ttnn/cpp/pybind11/device.hpp [hunk 10]
```diff
@@ -30,6 +31,37 @@ void py_module(py::module& module) {
 
     module.def("disable_and_clear_program_cache", &ttnn::disable_and_clear_program_cache, py::arg("device"), py::kw_only());
 
+    module.def("SetDefaultDevice", &ttnn::operations::experimental::auto_format::AutoFormat::SetDefaultDevice);
```
Comment (dmakoviichuk-tt) [inline ttnn/cpp/pybind11/device.hpp:34]: codestyle question. Should it be "set_default_device"?

**Review by sjameelTT — APPROVED** [review]

**Review by ntarafdar — APPROVED** [review]

File: conftest.py (hunk 9 above)
Comment (mouliraj-mcw) [reply conftest.py:117]: Hi @tt-rkim . It is handled in separate PR. Thankyou.

File: ttnn/cpp/pybind11/device.hpp (hunk 10 above)
Comment (mouliraj-mcw) [reply ttnn/cpp/pybind11/device.hpp:34]: Hi @dmakoviichuk-tt .
I followed the previous code style, but I'll confirm it to be sure. 
Thank you!

**Review by mywoodstock — APPROVED** [review]

File: ttnn/cpp/pybind11/device.hpp (hunk 10 above)
Comment (dmakoviichuk-tt) [reply ttnn/cpp/pybind11/device.hpp:34]: Please update PR description with a new added functions.

File: ttnn/cpp/ttnn/operations/experimental/auto_format/auto_format.hpp [hunk 11]
```diff
@@ -4,21 +4,21 @@
...
 #include "tt_metal/common/constants.hpp"
 #include "ttnn/operation.hpp"
 
 #include <optional>
 
 #include "tt_metal/common/math.hpp"
-using namespace tt::constants;
 
-namespace tt {
+using namespace tt::constants;
```
Comment (eyonland) [inline ttnn/cpp/ttnn/operations/experimental/auto_format/auto_format.hpp:16]: Please do not ever apply 'using nameaspace' in a header globally like this.

**Review by eyonland — CHANGES_REQUESTED** [review]
Do not use 'using namespace' globally in a header file.   Please move this inside of the functions or explicitly specify.

File: ttnn/cpp/pybind11/device.hpp (hunk 10 above)
Comment (mouliraj-mcw) [reply ttnn/cpp/pybind11/device.hpp:34]: Hi @dmakoviichuk-tt,
I didn't add any new functions, just migrated the existing ones.

**Review by dmakoviichuk-tt — APPROVED** [review]

**Comment by tt-rkim** [comment]
Can you please run:

- single card demos
- nightly fast dispatch (likely will fail, so should assert that no new errors are introduced)

**Review by eyonland — APPROVED** [review]
Approving on the condition that https://github.com/tenstorrent/tt-metal/issues/11989 is done the following day with PR ready to approve.

**Pull request merged**